fix(ssi): merge target and namespace implementations

betterengineering opened the pull request:
### What does this PR do?
When we introduced targets, we created a separate code path to handle the target based configuration. However, the namespace based configuration can also be represented and supported by the target implementation. This commit merges the two implementations.

When I merged these in code, I found bugs in the target implementation surrounding [Local SDK Injection](https://docs.datadoghq.com/tracing/guide/local_sdk_injection/) and how it was handled with targets defined. This change fixes those bugs to match the expectations around enabledNamespaces.

### Motivation
We want to reduce the complexity of the auto instrumentation webhook. There is no need to have two implementations to support our customer facing configuration options and as shown with this change, two implementations open up more chances for bugs.

See [SSI Kubernetes | Platform Stability](https://docs.google.com/document/d/1NqrPEUn3RfcdS_hQUQB-pJFJN9N-x5I202CB7s7CnwQ/edit?usp=sharing) for more details on all changes related to cleanup.

### Describe how you validated your changes
I tested this change heavily using [injector-dev](https://github.com/DataDog/injector-dev). For each scenario listed, I ran with and without the `--build` flag to use the latest agent release vs this branch to ensure there was a seamless migration path.

<details>
<summary>Broken annotation based injection with target list</summary>

This should have injection but it does not currently:

```yaml
helm:
  apps:
    - name: annotation-example
      namespace: application
      values:
        env:
          - name: DD_TRACE_DEBUG
            value: "true"
          - name: DD_APM_INSTRUMENTATION_DEBUG
            value: "true"
        image:
          repository: registry.ddbuild.io/ci/injector-dev/python
          tag: 2cd78ded
        podLabels:
          admission.datadoghq.com/enabled: "true"
          tags.datadoghq.com/env: local
        podAnnotations:
          admission.datadoghq.com/python-lib.version: "v3"
        service:
          port: "8080"
  versions:
    agent: 7.71.1
    cluster_agent:
      version: 7.71.1
      build: {}
    injector: 0.48.0
  config:
    datadog:
      apm:
        instrumentation:
          enabled: false
          targets:
            - name: python
              podSelector:
                matchLabels:
                  language: python
              ddTraceVersions:
                python: "3"
```

 </details>

<details>
<summary>Annotation based injection</summary>

We expect injection:
```yaml
helm:
  apps:
    - name: annotation-example
      namespace: application
      values:
        env:
          - name: DD_TRACE_DEBUG
            value: "true"
          - name: DD_APM_INSTRUMENTATION_DEBUG
            value: "true"
        image:
          repository: registry.ddbuild.io/ci/injector-dev/python
          tag: 2cd78ded
        podLabels:
          admission.datadoghq.com/enabled: "true"
          tags.datadoghq.com/env: local
        podAnnotations:
          admission.datadoghq.com/python-lib.version: "v3"
        service:
          port: "8080"
  versions:
    agent: 7.71.1
    cluster_agent:
      version: 7.71.1
      build: {}
    injector: 0.48.0
  config:
    datadog:
      apm:
        instrumentation:
          enabled: false
```

 </details>

<details>
<summary>Namespace based selection</summary>

We expect injection:
```yaml
helm:
  apps:
    - name: namespace-selection-example
      namespace: application
      values:
        env:
          - name: DD_TRACE_DEBUG
            value: "true"
          - name: DD_APM_INSTRUMENTATION_DEBUG
            value: "true"
        image:
          repository: registry.ddbuild.io/ci/injector-dev/python
          tag: 2cd78ded
        podLabels:
          tags.datadoghq.com/env: local
        service:
          port: "8080"
  versions:
    agent: 7.71.1
    cluster_agent:
      version: 7.71.1
      build: {}
    injector: 0.48.0
  config:
    datadog:
      apm:
        instrumentation:
          enabled: true
          enabledNamespaces:
            - "application"
          libVersions:
            python: "3"
```

 </details>

<details>
<summary>Workload selection</summary>

We expect injection:
```yaml
helm:
  apps:
    - name: workload-selection-example
      namespace: application
      values:
        env:
          - name: DD_TRACE_DEBUG
            value: "true"
          - name: DD_APM_INSTRUMENTATION_DEBUG
            value: "true"
        image:
          repository: registry.ddbuild.io/ci/injector-dev/python
          tag: 2cd78ded
        podLabels:
          language: python
          tags.datadoghq.com/env: local
        service:
          port: "8080"
  versions:
    agent: 7.71.1
    cluster_agent:
      version: 7.71.1
      build: {}
    injector: 0.48.0
  config:
    datadog:
      apm:
        instrumentation:
          enabled: true
          targets:
            - name: python
              podSelector:
                matchLabels:
                  language: python
              ddTraceVersions:
                python: "3"
```
 </details>

<details>
<summary>Instrumentation disabled</summary>

We expect no injection to occur:
```yaml
helm:
  apps:
    - name: disabled-example
      namespace: application
      values:
        env:
          - name: DD_TRACE_DEBUG
            value: "true"
          - name: DD_APM_INSTRUMENTATION_DEBUG
            value: "true"
        image:
          repository: registry.ddbuild.io/ci/injector-dev/python
          tag: 2cd78ded
        podLabels:
          tags.datadoghq.com/env: local
        service:
          port: "8080"
  versions:
    agent: 7.71.1
    cluster_agent:
      version: 7.71.1
      build: {}
    injector: 0.48.0
  config:
    datadog:
      apm:
        instrumentation:
          enabled: false
```
 </details>


### Additional Notes
This change opted to not change the behavior of any tests and to keep test changes minimal. A follow up PR will rewrite some of these tests to test the webhook behavior more directly.

betterengineering committed:
fix(ssi): merge target and namespace implementations

When we introduced targets, we created a seperate code path to handle
the target based configuration. However, the namespace based
configuration can also be represented and supported by the target
implementation. This commit merges the two.

betterengineering committed:
Fix linter

betterengineering committed:
Fix tests

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_mutator.go:
```diff
@@ -92,10 +105,16 @@ func NewTargetMutator(config *Config, wmeta workloadmeta.Component, imageResolve
+			pinnedLibraries := getPinnedLibraries(t.TracerVersions, config.containerRegistry, true)
+			if pinnedLibraries.areSetToDefaults {
+				usesDefaultLibs = true
+			}
```
```suggestion
			usesDefaultLibs = pinnedLibraries.areSetToDefaults
```

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_mutator.go:
```diff
@@ -178,6 +194,15 @@ func (m *TargetMutator) MutatePod(pod *corev1.Pod, ns string, _ dynamic.Interfac
+		if usingLanguageDetection {
+			extracted = extractedLanguageDetection
+		}
+
```
```suggestion
```

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_mutator.go:
```diff
@@ -353,6 +429,31 @@ func (t targetInternal) matchesPodSelector(podLabels map[string]string) bool {
+// we would inject all SDKs to all pods. This target encompasses both of those cases.
+func createDefaultTarget(namespaces []string, pinnedLibVersions map[string]string) Target {
+	// Create a default target.
+	target := Target{
```
Right now we add an annotation to the pod that says which target this is coming from and what the configuration is. 

Is that consistent for the behavior from namespaces? We'll have an `applied-target` with this JSON format?

What happens for the annotation? I'm not sure we should be adding an `applied-target` annotation there

betterengineering replied:
For `enabledNamespaces`, the generated target applies. I updated this change to no longer apply an annotation for local lib injection.

stanistan replied:
Thanks!

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_mutator_test.go:
```diff
@@ -679,6 +735,125 @@ func TestGetTargetLibraries(t *testing.T) {
 
+func TestLanguageDetection(t *testing.T) {
+	tests := map[string]struct {
+		config                     map[string]interface{}
```
nit: `any` instead of `interface{}`

stanistan commented on releasenotes-dca/notes/merge-namespace-targets-4c82b02d2c1dc1ca.yaml:
```diff
@@ -0,0 +1,16 @@
+  - |
+    Customers using Single Step Instrumentation with target based workload selection can now use language detection.
+    Language detection greatly reduces startup time when all default libraries are configured for a target. To enable
+    this feature, set `language_detection.enabled` to `true`.
```
`language_detection` is enabled by default so we don't need to add this here.

- https://github.com/DataDog/helm-charts/blob/main/charts/datadog/values.yaml#L593-L597
- https://github.com/DataDog/datadog-operator/blob/main/api/datadoghq/v2alpha1/datadogagent_types.go#L239-L245

stanistan commented on pkg/clusteragent/admission/controllers/webhook/controller_base.go:
```diff
@@ -189,7 +189,7 @@ func generateAutoInstrumentationWebhook(wmeta workloadmeta.Component, datadogCon
 	}
 
-	apm, err := autoinstrumentation.NewMutatorWithFilter(config, wmeta, imageResolver)
+	apm, err := autoinstrumentation.NewTargetMutator(config, wmeta, imageResolver)
```
One thing to consider is that we can have the NewMutatorWithFilter do everything that `NewTargetMutator` was doing (renaming that function) to keep the scope of this change to be _internal_ to the autoinstrumentation package and make it so that the webhook registration _outside_ of that stays exactly the same.

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_mutator.go:
```diff
@@ -215,10 +240,21 @@ func (m *TargetMutator) MutatePod(pod *corev1.Pod, ns string, _ dynamic.Interfac
 // a filter.
 func (m *TargetMutator) ShouldMutatePod(pod *corev1.Pod) bool {
-	// Return if the mutator is disabled.
+	enabledLabelVal, enabledLabelExists := getEnabledLabel(pod)
```
Add a comment clarifying that we are explicitly looking for "opt-out" of mutation here which is why we only check for the negation.

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_mutator.go:
```diff
@@ -263,37 +299,77 @@ type targetInternal struct {
+		return true
+	}
+
+	extractedLibraries := extractLibrariesFromAnnotations(pod, m.containerRegistry)
```
If I'm reading the PR correctly right now-- we're extracting the libraries for annotations multiple times now and there's a behavior between saying `true` here for `all` which affects the which libraries we're using. 

It's a bit confusing to do this work 2x and have the coupling across functions in a kind of circular way.

---

I think there's a way to clear this up:

1. [Don't build `internalTargets` if we ssi is not enabled](https://github.com/DataDog/datadog-agent/blob/bf19d6d246fb8390a5cc41d766ea8a5b5656d47c/pkg/clusteragent/admission/mutate/autoinstrumentation/target_mutator.go#L65-L155) -- this would then give us a list of _no targets_ (which I think is a source of issues because we compile and move on)
2. I think then we can remove [these checks](https://github.com/DataDog/datadog-agent/blob/bf19d6d246fb8390a5cc41d766ea8a5b5656d47c/pkg/clusteragent/admission/mutate/autoinstrumentation/target_mutator.go#L248-L255) from `ShouldMutatePod()` since targets are empty and the existence of a target implies if it will be mutated or not
3. The only target that _might_ be there after this is if we're doing annotations which we already check in `getTarget()` 
4. Because of the clean up in (2), we won't need multiple calls to `extractLibraries...` and can drop `podHasLibraryAnnotations` and can move the `all` check to be explicit in `getTargetFromAnnotation`

stanistan replied:
Trying this out here here:

- https://github.com/DataDog/datadog-agent/pull/42200

stanistan reviewed (commented):
Great cleanup! Have a few suggestions/questions on the implementation wrt annotations/target simplification

betterengineering committed:
Apply suggestion from @stanistan

Co-authored-by: Stan Rozenraukh <stan.rozenraukh@datadoghq.com>

betterengineering committed:
Apply suggestion from @stanistan

Co-authored-by: Stan Rozenraukh <stan.rozenraukh@datadoghq.com>

betterengineering committed:
Apply suggestion from @stanistan

Co-authored-by: Stan Rozenraukh <stan.rozenraukh@datadoghq.com>

betterengineering committed:
Add comment on should mutate pod

betterengineering committed:
Update release note

betterengineering committed:
Use any

betterengineering committed:
Fixup applied target annotation

stanistan committed:
autoinstrumentation: update annotations targets to not do work twice (#42200)

### What does this PR do?

Attempting the small cleanup for #42068 

### Motivation

### Describe how you validated your changes

### Additional Notes

buraizu commented on releasenotes-dca/notes/merge-namespace-targets-4c82b02d2c1dc1ca.yaml:
```diff
@@ -0,0 +1,15 @@
+---
+features:
+  - |
+    Customers using Single Step Instrumentation with target based workload selection can now use language detection.
```
```suggestion
    Customers using Single Step Instrumentation with target-based workload selection can now use language detection.
```

buraizu commented on releasenotes-dca/notes/merge-namespace-targets-4c82b02d2c1dc1ca.yaml:
```diff
@@ -0,0 +1,15 @@
+fixes:
+  - |
+    There were several bugs for customers using Single Step Instrumentation with target based workload selection and
+    also utilized Local SDK Injection. This change resolves an issue where if targets were defined, the cluster agent
```
```suggestion
    also using local SDK injection. This change resolves an issue where if targets were defined, the Cluster Agent
```

buraizu commented on releasenotes-dca/notes/merge-namespace-targets-4c82b02d2c1dc1ca.yaml:
```diff
@@ -0,0 +1,15 @@
+    There were several bugs for customers using Single Step Instrumentation with target based workload selection and
+    also utilized Local SDK Injection. This change resolves an issue where if targets were defined, the cluster agent
+    didn't respect the `admission.datadoghq.com/enabled` annotation or the `admission_controller.mutate_unlabelled`
+    configuration option and only respected the language annotations.
```
```suggestion
    configuration option, and only respected the language annotations.
```

buraizu reviewed (changes requested):
Requesting a few updates for branding, style, and clarity

betterengineering committed:
Apply suggestions from code review

Co-authored-by: Bryce Eadie <bryce.eadie@datadoghq.com>
